pydy/tests/test_joints.py

```diff
@@ -0,0 +1,275 @@
+        assert cross(self.pinjoint.child_axis, self.pinjoint.parent_axis) == - self.parent.frame.z
+
+        self.align_axes()
+        parent_axis = self.pinjoint.parent_axis
+        child_axis = self.pinjoint.child_axis
+        assert acos(dot(parent_axis, child_axis)/(parent_axis.magnitude() * child_axis.magnitude())) == 0
+
+        # Similar to Pinjoint but have a velocity instead of angular velocity.
+        assert self.child_point.vel(self.parent.frame) != 0
+
+class TestCylindricalJoint():
+    # Note all the functionalitites are similar to other PinJoint and SlidingJoint.Only CylindricalJoint's specific
+    # tests are written.
+    def setup(self):
+        self.parent = Body('parent')
+        self.child = Body('child')
+        self.cylindricaljoint = CylindricalJoint('cylindricaljoint', self.parent, self.child, par_point_vec_tuple=(1,1,1),
+                                                 child_point_vec_tuple=(0,0,1), parent_direction='x + 2*y', child_direction = 'z')
+                                                 #Similar approach can be used in other joints as well.
+
+    def test_cylindricaljoint_parameters_assignment(self):
+        assert self.cylindricaljoint.parent == self.parent
+        assert self.cylindricaljoint.child == self.child
+        assert self.cylindricaljoint.parent_joint_vector == self.parent.frame.x + self.parent.frame.y + self.parent.frame.z
+        assert self.cylindricaljoint.child_joint_vector == self.child.frame.z
+        assert self.cylindricaljoint.parent_direction == self.parent.frame.x + 2*self.parent.frame.y
+        assert self.cylindricaljoint.child_direction == self.child.frame.z
+
+    def test_cylindricaljoint_apply_joint(self):
+        assert self.child_point.vel(self.parent.frame) != 0
+        parent_axis = self.pinjoint.parent_direction
+        child_axis = self.pinjoint.child_direction
+        assert acos(dot(parent_axis, child_axis)/(parent_axis.magnitude() * child_axis.magnitude())) == 0
+
+
+class TestSphericalJoint():
+    def setup(self):
+        self.parent = Body('parent')
+        self.child = Body('child')
+        self.sphericaljoint = SphericalJoint(self.parent, self.child, par_point_vec_tuple=(0,2,0),
+                                             child_point_vec_tuple=(0,1,1), parent_plane='x*y', child_plane='y*z + 2*z*x')
+                                             # yz is y-z plane
+        # Note: Refer to SphericalJoint's docstring for information about implementation using planes.
+
+    def test_spehricaljoint_paramters_assignment(self):
+        assert self.sphericaljoint.parent == self.parent
+        assert self.sphericaljoint.child == self.child
+        assert self.sphericaljoint.parent_joint_vector == 2 * self.parent.frame.y
+        assert self.sphericaljoint.child_joint_vector == self.child.frame.y + self.child.frame.z
+        assert self.sphericaljoint.parent_plane == # TODO
+        assert self.sphericaljoint.child_plane == # TODO
```

Comment:
This test is missing all info about coordinates and velocities. It should use quaternions as the base implementation. So there will be four coordinates and four velocities. Try making a spherical joint manually with quaternions and then write the tests that will exercise that.